src/modules-load/modules-load.c

```diff
@@ -146,39 +398,82 @@ static int run(int argc, char *argv[]) {
                 log_warning_errno(r, "Failed to parse kernel command line, ignoring: %m");
 
-        r = module_setup_context(&ctx);
-        if (r < 0)
-                return log_error_errno(r, "Failed to initialize libkmod context: %m");
-
-        r = 0;
-
         if (argc > optind) {
-                for (int i = optind; i < argc; i++)
-                        RET_GATHER(r, apply_file(ctx, argv[i], false));
-
+                for (int i = optind; i < argc; i++) {
+                        r = apply_file_from_path(argv[i], module_set);
+                        if (r < 0) {
+                                RET_GATHER(ret, r);
+                                break;
+                        }
+                }
         } else {
-                _cleanup_strv_free_ char **files = NULL;
+                ConfFile **files = NULL;
+                size_t n_files = 0;
+
+                CLEANUP_ARRAY(files, n_files, conf_file_free_many);
 
                 STRV_FOREACH(i, arg_proc_cmdline_modules) {
-                        k = module_load_and_warn(ctx, *i, true);
-                        if (k == -ENOENT)
-                                continue;
-                        RET_GATHER(r, k);
+                        r = modules_list_append_dup(module_set, *i);
+                        if (r < 0) {
+                                log_error_errno(r, "Failed to add module '%s': %m", *i);
+                                RET_GATHER(ret, r);
+                        }
                 }
 
-                k = conf_files_list_nulstr(&files, ".conf", NULL, 0, conf_file_dirs);
-                if (k < 0)
-                        return log_error_errno(k, "Failed to enumerate modules-load.d files: %m");
+                r = conf_files_list_nulstr_full(".conf", /* root = */ NULL,
+                                                CONF_FILES_REGULAR | CONF_FILES_FILTER_MASKED,
+                                                conf_file_dirs, &files, &n_files);
+                if (r < 0)
+                        RET_GATHER(ret, log_error_errno(r, "Failed to enumerate modules-load.d files: %m"));
+                else
+                        FOREACH_ARRAY(cf, files, n_files)
+                                RET_GATHER(ret, apply_conf_file(*cf, module_set));
+        }
+
+        num_threads = determine_num_worker_threads(ordered_set_size(module_set));
 
-                STRV_FOREACH(fn, files)
-                        RET_GATHER(r, apply_file(ctx, *fn, true));
+        /* If no additional thread is required, there is no need to create the
+         * thread pool or the mean to communicate with its members. */
+        if (num_threads == 0) {
+                log_debug("Single-threaded probe");
+                return RET_GATHER(ret, do_direct_probe(module_set));
         }
 
-        return r;
+        /* Create a socketpair for communication with probe workers */
+        r = RET_NERRNO(socketpair(AF_UNIX, SOCK_SEQPACKET | SOCK_CLOEXEC, 0, pair));
```

Comment:
SOCK_DGRAM should suffice, you don't need bi-directional operation

Reply:
A socket pair using `SOCK_DRGAM` does not return 0 in `recv()` when the sending end (i.e.: `pair[0]`) is closed, so `SOCK_SEQPACKET` is required here.